### PR TITLE
Improve focus behaviour in search

### PR DESCRIPTION
**Pull request by krassowski**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/14308

Depends on https://github.com/codemirror/view/commit/f49f2eea1ae1b81c8e943b76a57ca787bfd46bf5 included in [@codemirror/view](https://github.com/codemirror/view) 6.9.5, therefore required https://github.com/jupyterlab/jupyterlab/pull/14413

## Code changes

- only create DOM event (i.e. focus) CM extension once, and only add it during editor setup phase
- support arbitrary nesting of highlight markers fixing current highlight not shown in some cases
- preserve current match when switching cells if possible 
- if there is no current match, try to assign one
- do not ever set selection when searching in selection, instead just scroll to the match
- introduces (non-public) `SearchStartAnchor` type to disambiguate the search start behaviour allowing to fix highlighting issues
- introduces `IHighlightAdjacentMatchOptions` interface to specify whether highlighting should scroll/select new match
- do not scroll/select new match when starting search
   - this can be improved in follow-up: we could scroll/select if this is a new query (but not restart of query after document change)

## User-facing changes

Extending selection after changing to command mode now works smoothly (does not revert to edit mode, which was caused by CodeMirror bug, and is faster - which was previously slowed down by needless cleaning of highlights due to both `activeCellChanged` and `selectionCellChanged` handlers being used - now the shared job will only be performed once).

Initial match is set consistently across various scenarios, e.g.:

| Before | After |
|---|--|
| image | image |

Including, in file editor, and when populating search box from selection (regardless of selection direction):

image

## Backwards-incompatible changes

N/A

### CODE REVIEW

File: packages/codemirror/style/base.css [hunk 1]
```diff
@@ -59,24 +59,27 @@
...
 .cm-searching,
 .cm-searching span {
+  /* `.cm-searching span`: we need to override syntax highlighting */
   background-color: var(--jp-search-unselected-match-background-color);
   color: var(--jp-search-unselected-match-color);
 }
 
-.cm-content .cm-searching::selection,
-.cm-content .cm-searching span::selection {
+.cm-searching::selection {
```
Comment (krassowski) [inline packages/codemirror/style/base.css:67]:
```suggestion
.cm-searching::selection,
.cm-searching span::selection {
```

**Comment by krassowski** [comment]
Merged master to clean update-links job failure that was addressed earlier.

**Review by fcollonval — COMMENTED** [review]
Thanks a lot @krassowski for working on this.

**Review by fcollonval — APPROVED** [review]

**Comment by krassowski** [comment]
@andrii-i can we merge this one or did you have any specific feedback here?

**Comment by fcollonval** [comment]
Merging now to get it in the RC

**Comment by andrii-i** [comment]
@krassowski No feedback on my end, it was mentioned as needing a review last week but I have not came around to reviewing. Thanks for reviewing @fcollonval.

**Pull request merged**